### PR TITLE
Make crop work the same for pil and tensor

### PR DESCRIPTION
**Pull request by Jaesuny**
In the `transforms.functional.crop`, if the cropping area is out of the image, in the case of pil, the box will be padded by zeros, whereas in the tensor, only the inside of the image will be returned.

See this [notebook](https://colab.research.google.com/drive/1r8SOPEF1Zv54WD-R2f_xLGOTHicmdH5N?usp=sharing).

In this PR `functional_tensor.crop` pad if needed.

### CODE REVIEW

File: torchvision/transforms/functional_tensor.py [hunk 1]
```diff
@@ -122,7 +122,12 @@ def hflip(img: Tensor) -> Tensor:
 def crop(img: Tensor, top: int, left: int, height: int, width: int) -> Tensor:
     _assert_image_tensor(img)
 
-    return img[..., top:top + height, left:left + width]
+    w, h = _get_image_size(img)
+    right = left + width
+    bottom = top + height
+
+    padding = [max(-left, 0), max(-top, 0), max(right - w, 0), max(bottom - h, 0)]
+    return pad(img[..., top:top + height, left:left + width], padding)
```
Comment (vfdev-5) [inline torchvision/transforms/functional_tensor.py:130]: Let's only call `pad` if needed like here: https://github.com/pytorch/vision/pull/3333/files

**Review by vfdev-5 — COMMENTED** [review]
Thanks for the PR @Jaesuny , I think we can avoid calling pad all the times. Let's improve the code like that.
I think documentation update would be also needed ...

**Comment by Jaesuny** [comment]
> Thanks for the PR @Jaesuny , I think we can avoid calling pad all the times. Let's improve the code like that.
> I think documentation update would be also needed ...

Thanks for the review @vfdev-5 . Updated based on your feedback.

File: torchvision/transforms/functional_tensor.py [hunk 2]
```diff
@@ -122,7 +122,14 @@ def hflip(img: Tensor) -> Tensor:
 def crop(img: Tensor, top: int, left: int, height: int, width: int) -> Tensor:
     _assert_image_tensor(img)
 
-    return img[..., top:top + height, left:left + width]
+    w, h = _get_image_size(img)
+    right = left + width
+    bottom = top + height
+
+    if left < 0 or top < 0 or right > w or bottom < h:
```
Comment (fmassa) [inline torchvision/transforms/functional_tensor.py:129]: Is this condition for `bottom` correct? I would have imagined that it should be similar than the one for `right` (which uses a `>` instead of a `<`).

File: test/test_transforms_tensor.py [hunk 3]
```diff
@@ -188,6 +188,10 @@ def test_crop(self):
             'crop', 'RandomCrop', fn_kwargs=fn_kwargs, meth_kwargs=meth_kwargs
         )
 
+        # Test transforms.functional.crop including outside the image area
+        fn_kwargs = {"top": -2, "left": 8, "height": 4, "width": 5}
```
Comment (fmassa) [inline test/test_transforms_tensor.py:192]: Can you add more tests exercising other use-cases, like when `top` is negative and `height + top` is larger than the image size?

**Review by fmassa — COMMENTED** [review]
Thanks for the PR!

I have a few comments, let me know what you think

File: torchvision/transforms/functional_tensor.py (hunk 2 above)
Comment (Jaesuny) [reply torchvision/transforms/functional_tensor.py]: I missed it.
Fixed this.
Thank you!

File: test/test_transforms_tensor.py (hunk 3 above)
Comment (Jaesuny) [reply test/test_transforms_tensor.py]: Added tests considering all directions!

**Comment by fmassa** [comment]
Thanks for the changes!

Lint is failing in https://app.circleci.com/pipelines/github/pytorch/vision/8117/workflows/29ea8926-e7f7-448b-b705-53905de1de37/jobs/580130 , but once you fix it the PR is good to merge.

File: test/test_transforms_tensor.py [hunk 4]
```diff
@@ -188,6 +188,22 @@ def test_crop(self):
             'crop', 'RandomCrop', fn_kwargs=fn_kwargs, meth_kwargs=meth_kwargs
         )
 
+        # Test transforms.functional.crop including outside the image area
+        fn_kwargs = {"top": -2, "left": 3, "height": 4, "width": 5} # top
```
Comment (datumbox) [inline test/test_transforms_tensor.py:192]:
```suggestion
        fn_kwargs = {"top": -2, "left": 3, "height": 4, "width": 5} #  top
```

File: test/test_transforms_tensor.py [hunk 5]
```diff
@@ -188,6 +188,22 @@ def test_crop(self):
...
+        self._test_functional_op('crop', fn_kwargs=fn_kwargs)
+
+        fn_kwargs = {"top": 1, "left": -3, "height": 4, "width": 5} # left
```
Comment (datumbox) [inline test/test_transforms_tensor.py:195]:
```suggestion
        fn_kwargs = {"top": 1, "left": -3, "height": 4, "width": 5} #  left
```

File: test/test_transforms_tensor.py [hunk 6]
```diff
@@ -188,6 +188,22 @@ def test_crop(self):
...
+        self._test_functional_op('crop', fn_kwargs=fn_kwargs)
+
+        fn_kwargs = {"top": 7, "left": 3, "height": 4, "width": 5} # bottom
```
Comment (datumbox) [inline test/test_transforms_tensor.py:198]:
```suggestion
        fn_kwargs = {"top": 7, "left": 3, "height": 4, "width": 5} #  bottom
```

File: test/test_transforms_tensor.py [hunk 7]
```diff
@@ -188,6 +188,22 @@ def test_crop(self):
...
+        self._test_functional_op('crop', fn_kwargs=fn_kwargs)
+
+        fn_kwargs = {"top": 3, "left": 8, "height": 4, "width": 5} # right
```
Comment (datumbox) [inline test/test_transforms_tensor.py:201]:
```suggestion
        fn_kwargs = {"top": 3, "left": 8, "height": 4, "width": 5} #  right
```

File: test/test_transforms_tensor.py [hunk 8]
```diff
@@ -188,6 +188,22 @@ def test_crop(self):
...
+        self._test_functional_op('crop', fn_kwargs=fn_kwargs)
+
+        fn_kwargs = {"top": -3, "left": -3, "height": 15, "width": 15} # all
```
Comment (datumbox) [inline test/test_transforms_tensor.py:204]:
```suggestion
        fn_kwargs = {"top": -3, "left": -3, "height": 15, "width": 15} #  all
```

**Review by datumbox — COMMENTED** [review]
Appeasing the formatter gods.

**Review by datumbox — APPROVED** [review]
